12-monads/Risk.hs

```diff
@@ -0,0 +1,172 @@
+ battle :: Battlefield -> Rand StdGen Battlefield
+which simulates a single battle (as explained above) between two opposing armies. That is, it should simulate randomly rolling the appropriate number of dice, interpreting the results, and updating the two armies to reflect casualties. You may assume that each player will attack or defend with the maximum number of units they are allowed.
+-}
+--Exercise 2
+battle :: Battlefield -> Rand StdGen Battlefield
+battle bf@(Battlefield att def) = do
+    attDice <- sortedDiceVals maxAtt
+    defDice <- sortedDiceVals maxDef
+    let didWin = zipWith (>) attDice defDice
+
+    return $ foldr updateBf bf didWin
+      where
+        maxAtt :: Int
+        maxAtt = min 3 (att - 1)
+
+        maxDef :: Int
+        maxDef = min 2 def
+
+        sortedDiceVals :: Int -> Rand StdGen [Int]
+        sortedDiceVals n = map unDV . reverse . sort <$> replicateM n die
+
+        updateBf :: Bool -> Battlefield -> Battlefield
+        updateBf win (Battlefield att def)
+            | win       = Battlefield att (def - 1)
+            | otherwise = Battlefield (att - 1) def
+
+
+{-
+Exercise 3
+Of course, usually an attacker does not stop after just a single
+battle, but attacks repeatedly in an attempt to destroy the entire de- fending army (and thus take over its territory).
+Now implement a function
+invade :: Battlefield -> Rand StdGen Battlefield
+which simulates an entire invasion attempt, that is, repeated calls to battle until there are no defenders remaining, or fewer than two attackers.
+-}
+--Exercise 3
+invade :: Battlefield -> Rand StdGen Battlefield
+invade bf@(Battlefield att def)
+    | att < 2 || def <= 0 = return bf
+    | otherwise           = battle bf >>= invade
+
+{-
+Exercise 4
+Finally, implement a function
+ successProb :: Battlefield -> Rand StdGen Double
+which runs invade 1000 times, and uses the results to compute a Double between 0 and 1 representing the estimated probability that the attacking army will completely destroy the defending army.
+For example, if the defending army is destroyed in 300 of the 1000 simulations (but the attacking army is reduced to 1 unit in the other 700), successProb should return 0.3.
+-}
+--Exercise 4
+successProb :: Battlefield -> Rand StdGen Double
+successProb bf = do
+    res <- replicateM 1000 (invade bf)
+    let winCount = length $ filter id (fmap isWin res)
+    return $ fromIntegral winCount / 1000
+      where
+        isWin :: Battlefield -> Bool
+        isWin (Battlefield att def) = att > def
+
+{-
+Exercise 5 (Optional) Write a function
+ exactSuccessProb :: Battlefield -> Double
+which computes the exact probability of success based on principles of probability, without running any simulations. (This won’t give you any particular practice with Haskell; it’s just a potentially interesting challenge in probability theory.)
+-}
+--Exercise 5
+exactSuccessProb :: Battlefield -> Double
+exactSuccessProb bf = evalRand (successProb bf) (mkStdGen 1)
```

Comment:
This is not calculating exact probability, just running the simulation once with fixed field.
Simulating won't help you solve this. You'll have to find a way to calculate the probability of winning mathematically.

Reply:
@jhrcek Okay, I think I will leave this as a handy math challenge when I have some free time 😂 
https://blog.wolfram.com/2017/11/20/how-to-win-at-risk-exact-probabilities/